### PR TITLE
Remove obsolete resolving logic from TRA

### LINKED ISSUE #20279: TRA waits when an index doesn't exist but fails immediately when shard is not found
TransportReplicationAction currently has an inconsistency in how it deals with requests that refer to things that don't exist (which is different than not available).

1) When an index is not found in the cluster state, we go into a retry loop where we wait for the index to appear.
2) When a request comes in for a shard that doesn't exists (i.e., the shard id is higher than the number of shards ) we fail immediately - as it will never appear.

This is surprising and we should fix it.

In my opinion we should:
1) Require ReplicationRequests to have a complete ShardId when they get to the reroute phase in TRA.
2) Fail immediately when that shard id can not be resolved.
3) Change TransportIndexAction and similar write actions to resolve the incoming requests and set their proper shard id (with index uuid). If they need to create the index, they can go ahead, but then it's up to them to also wait until the current (data) node, knows about the index that was just created. We can have a shared utility method for this on `AutoCreateIndex`.

### PR DESCRIPTION
**Pull request by ywelsch**
This stems from a time where index requests were directly forwarded to TransportReplicationAction. Nowadays they are wrapped in a BulkShardRequest, and this logic is obsolete.

Closes #20279

### CODE REVIEW

**Review by original-brownbear — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 1]
```diff
@@ -219,21 +219,10 @@ public ClusterBlockLevel indexBlockLevel() {
...
-    protected boolean resolveIndex() {
-        return true;
-    }
-
     protected TransportRequestOptions transportOptions(Settings settings) {
         return TransportRequestOptions.EMPTY;
     }
 
-    private String concreteIndex(final ClusterState state, final ReplicationRequest request) {
-        return resolveIndex() ? indexNameExpressionResolver.concreteSingleIndex(state, request).getName() : request.index();
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: I think we can remove `IndexNameExpressionResolver` from this class and its subclasses.

**Review by dnhatn — APPROVED** [review]
I left a minor comment; LGTM.

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: Good point, removed in aa42739

**Comment by ywelsch** [comment]
I had to revert this PR, as it was making CI flaky. The reason is that the index expression logic had an interesting side effect. For bulk requests, the reroute phase was waiting for the index to appear in case where it was not present, and for all other replication requests it would right away throw an `IndexNotFoundException` and exit. With this change here, every replication request was now waiting for the index to appear, which was problematic when the given index had just been deleted (e.g. deleting a follower index while it's still receiving requests from the leader, where these requests would now wait up to a minute for the index to appear). I will follow-up with a change that takes this into consideration.

**Pull request merged**